sstables/sstable_directory.cc

```diff
@@ -122,6 +124,11 @@ sstable_directory::process_descriptor(sstables::entry_descriptor desc, const ::i
             return make_ready_future<>();
         }
     }).then([sst, this] {
+        if (_load_and_stream) {
```

Comment:
This yet-another-boolean-flag is ugly. How about changing process_descriptor() to always collect sstables in a vector, and then adding an extra pass to do the per-shard sorting in a new function, which the existing users would call?

Then the shard-sorting is opt-in by whoever wants it.

Reply:
I am not confident about the changes in sstables and distributed_loader.  I will think about it.

Reply:
I tried. sstable_directory::process_descriptor only processes one sstable. So, we need to pass a flag to distributed_loader::process_sstable_dir and sstable_directory::process_sstable_dir everywhere to instruct to call one of or both of sstable_directory::process_descriptor_collect_only() and sstable_directory::process_descriptor_sort_only().

I think it is more complicated and even more ugly than this version.

Reply:
Think of it from the point of view of someone maintaining sstable_directory. What does `sstables::sstable_directory::load_and_stream::yes` mean? It doesn't mean that it will load and stream sstables. It can only be understood by understanding its caller.

It's bad design for a function to behave differently depending on the name of the caller (it's okay to have flags, but the flag can't mean "called from X".

My suggestion is to change process_descriptor to avoid the complicated sorting it does not:

```
future<>
sstable_directory::process_descriptor(sstables::entry_descriptor desc, const ::io_priority_class& iop) {
    if (desc.version > _max_version_seen) {
        _max_version_seen = desc.version;
    }

    auto sst = _sstable_object_from_existing_sstable(_sstable_dir, desc.generation, desc.version, desc.format);
    return sst->load(iop).then([this, sst] {
        validate(sst);
        if (_need_mutate_level) {
            dirlog.trace("Mutating {} to level 0\n", sst->get_filename());
            return sst->mutate_sstable_level(0);
        } else {
            return make_ready_future<>();
        }
    }).then([sst, this] {
        return sst->get_open_info().then([sst, this] (sstables::foreign_sstable_open_info info) {
            auto shards = sst->get_shards_for_this_sstable();
            if (shards.size() == 1) {
                if (shards[0] == this_shard_id()) {
                    dirlog.trace("{} identified as a local unshared SSTable", sst->get_filename());
                    _unshared_local_sstables.push_back(sst);
                } else {
                    dirlog.trace("{} identified as a remote unshared SSTable", sst->get_filename());
                    _unshared_remote_sstables[shards[0]].push_back(std::move(info));
                }
            } else {
                dirlog.trace("{} identified as a shared SSTable", sst->get_filename());
                _shared_sstable_info.push_back(std::move(info));
            }
            return make_ready_future<>();
        });
    });
}
```

We can drop the second continuation and replace it with something like:

```c++
    }).then([sst, this] {
        _unsorted_sstables.push_back(sst);
    });
```

And then call a new function (or add a new flag sort_sstables_according_to_owner) to pick up _unsorted_sstables and apply the shard sorting step.

Reply:
> Think of it from the point of view of someone maintaining sstable_directory. What does sstables::sstable_directory::load_and_stream::yes mean? It doesn't mean that it will load and stream sstables. It can only be understood by understanding its caller.

I agree it is hard to figure out what does this flag mean.  It is better to name the flag sstables::sstable_directory::collect_sstables_only::yes.

Reply:
> My suggestion is to change process_descriptor to avoid the complicated sorting it does not:

I think I got your idea to split the sort part out, but this means we need to carry a flag everywhere in other places.

Reply:
@avikivity I pushed v3 to split out the "sort sstables" as you suggested. Let's make some progress.